### PR TITLE
PMP isotropic_remeshing() - improve edge flip step

### LINKED ISSUE #4144: Isotropic remeshing creates self intersections
## Issue Details

* Load the mesh in the Polyhedron demo
* Select 60 degree edges
* Isotropic remesh :  0.0015,  3 iterations, 3 smoothings, protect, allow

[cheese_transformed-facets.zip](https://github.com/CGAL/cgal/files/3485856/cheese_transformed-facets.zip)

The resulting mesh has self intersections as well as triangles with very small angles

## Environment

* CGAL version: master

### PR DESCRIPTION
**Pull request by janetournois**
## Summary of Changes

This PR improves the edge flipping step of `PMP::isotropic_remeshing()` by adding another criterion to trigger a flip : if one of the incident faces is badly shaped, the flip is tried, independently from the valence deviation improvement.

## Release Management

* Affected package(s): PMP
* Issue(s) solved (if any): fix #4144 
* License and copyright ownership: unchanged

this PR replaces #4192

### CODE REVIEW

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/self_intersections.h [hunk 1]
```diff
@@ -238,6 +238,9 @@ self_intersections_impl(const FaceRange& face_range,
   // This loop is very cheap, so there is hardly anything to gain from parallelizing it
   for(face_descriptor f : face_range)
   {
+    if (f == boost::graph_traits<TriangleMesh>::null_face())
```
Comment (sloriot) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/self_intersections.h:241]: how can this be possible? are you calling the function from a range of halfedge, some of them being on the border?

Comment (janetournois) [reply Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/self_intersections.h:241]: this commit was part of the PR #4192
I am pretty sure I had met this case in the test case I was running at that time...
I cost of that check is low and prevents for misusing the function, so I think we can keep it

Comment (sloriot) [reply Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/self_intersections.h:241]: @MaelRL what do you think? I'm not in favor of adding it and would prefer fixing the calling code.

Comment (MaelRL) [reply Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/self_intersections.h:241]: I also would prefer if this function assumed a clean range.

Comment (lrineau) [reply Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/self_intersections.h:241]: On the other hand, being robust to users' errors is also a nice property, no?

Comment (sloriot) [reply Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/self_intersections.h:241]: Well, I usually agree but here we are talking about a NULL element in an input range.

Comment (lrineau) [reply Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/self_intersections.h:241]: So the actual bug is in the calling code, from `demo/Polyhedron/`, right?

**Comment by janetournois** [comment]
ok let's assume the user is not giving the null face in the input range

File: Polygon_mesh_processing/test/Polygon_mesh_processing/remeshing_test.cpp [hunk 2]
```diff
@@ -171,74 +173,130 @@ Main(int argc, char* argv[])
...
+    t.stop();
+    std::cout << "Remeshing all took " << t.time() << std::endl;
+
+  }
+  else if (sharp_angle > 0)
+  {
+    typedef boost::property_map<Mesh, CGAL::edge_is_feature_t>::type EIFMap;
+    typedef boost::property_map<Mesh, CGAL::face_patch_id_t<int> >::type PIMap;
+    typedef boost::property_map<Mesh, CGAL::vertex_incident_patches_t<int> >::type VIMap;
+
```
Comment (sloriot) [inline Polygon_mesh_processing/test/Polygon_mesh_processing/remeshing_test.cpp:254]:
```suggestion
    typedef typename  boost::property_map<Mesh, CGAL::edge_is_feature_t>::type EIFMap;
    typedef typename boost::property_map<Mesh, CGAL::face_patch_id_t<int> >::type PIMap;
    typedef typename boost::property_map<Mesh, CGAL::vertex_incident_patches_t<int> >::type VIMap;

```

**Comment by maxGimeno** [comment]
Runtime errors : https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-5.3-Ic-76/Surface_mesh_approximation/TestReport_lrineau_Debian-stable-Release.gz

**Comment by sloriot** [comment]
@janetournois are you still targeting 5.3?

**Comment by janetournois** [comment]
Yes. I am working on it

**Comment by sloriot** [comment]
Should we disable that particular test while we debug what is wrong in the reported issue? It has nothing to do with the current patch of Jane.

**Comment by maxGimeno** [comment]
There are conflicts with master

**Comment by maxGimeno** [comment]
Successfully tested in https://cgal.geometryfactory.com/CGAL/testsuite/results-5.3-Ic-106.shtml

**Pull request merged**